### ISSUE TITLE
the loss is not correct

### ISSUE THREAD

**Issue by goodfeli** [carpedm20/DCGAN-tensorflow#15]
In this block:

        self.d_loss_real = binary_cross_entropy_with_logits(tf.ones_like(self.D), self.D)
        self.d_loss_fake = binary_cross_entropy_with_logits(tf.zeros_like(self.D_), self.D_)
        self.g_loss = binary_cross_entropy_with_logits(tf.ones_like(self.D_), self.D_)

You are using self.D and self.D_ as the logits, but they are probabilities, not the logits. The logits are the value that come before the sigmoid.

Your binary_cross_entropy_with_logits function is just entirely wrong and should probably be deleted. You should use tf.nn.sigmoid_cross_entropy_with_logits instead.

**Comment by carpedm20** [comment]
I think the variable names are wrong (logits and target should be change) but the calculation seems to be right for me.

In [tensorflow document](https://www.tensorflow.org/versions/r0.8/api_docs/python/nn.html#sigmoid_cross_entropy_with_logits), `tf.nn.sigmoid_cross_entropy_with_logits` is calculated as:

```
let x = logits, z = targets.
z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x))
```

and in my understanding, it wants to calculate:

```
p_target * -log(p_pred) + (1 - p_target) * -log(1 - sigmoid(p_pred)).
```

`self.D` is `tf.nn.sigmoid(h4)` like [this](https://github.com/carpedm20/DCGAN-tensorflow/blob/master/model.py#L195) which makes it to be p_pred and `binary_cross_entropy_with_logits` is calculated as:

```
tf.reduce_mean(-(logits * tf.log(targets) + (1. - logits) * tf.log(1. - targets)))
= tf.reduce_mean(-(tf.ones_like(self.D) * tf.log(self.D) + (1. - tf.ones_like(self.D)) * tf.log(1. - self.D)))
```

and I think this is right because `tf.ones_like(self.D)` is a target probability and `self.D` is a predicted probability which will calculate cross entropy of them.

As your suggestion, if we use:

```
self.d_loss_real = tf.nn.sigmoid_cross_entropy_with_logits(self.D, tf.ones_like(self.D)),
```

`self.D` (probability) will be passed into `sigmoid(x)` and the calculation will looks like:

```
z * -log(sigmoid(self.D)) + (1 - z) * -log(1 - sigmoid(self.D))
= p_target * -log(sigmoid(p_pred)) + (1 - p_target) * -log(1 - sigmoid(p_pred))
```

and this seems to be wrong way to calculate cross entropy.

I'm not an expert like you so please give me an advice if there is anything wrong in my explanation.

**Comment by goodfeli** [comment]
When computing the loss, you should not use preds at all. Preds involves a sigmoid so it can saturate and fail to learn. That is why your implementation has a bug where if you set the learning rate slightly too high, it will get stuck printing the same loss value (of roughly 22) over and over again. A correctly implemented cross entropy cost cannot get stuck at a specific high value.

My suggestion is not to use:

    self.d_loss_real = tf.nn.sigmoid_cross_entropy_with_logits(self.D, tf.ones_like(self.D))

My suggestion is to use:

   self.d_loss_real = tf.nn.sigmoid_cross_entropy_with_logits(self.D_logits, tf.ones_like(self.D)),

where

   self.D = tf.sigmoid(self.D_logits).

Your cross entropy calculation is incorrect because you compute the preds first, and the sigmoid can get numerically rounded to 0 or 1. To compensate for that, you have to do tricks like adding and subtracting eps:
https://github.com/carpedm20/DCGAN-tensorflow/blob/master/ops.py#L35
to avoid taking the log of zero, etc.
If implemented correctly, the cross-entropy does not need any eps tricks. The tricks make your cross entropy calculation only slightly inaccurate for most inputs. The important thing is that they make the gradient of the cross entropy very inaccurate for extreme inputs.
tf.nn.sigmoid_cross_entropy_with_logits has the correct implementation.

There's a longer explanation of cross-entropy and why it can't get stuck if implemented properly in section 6.2.1 of the textbook here: http://www.deeplearningbook.org/contents/mlp.html

**Comment by carpedm20** [comment]
@goodfeli Thanks for nice explanation! I think I need to fix other models too.
